src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -208,6 +208,108 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
+                        // locations such as anonymous types, etc...
+                        if (location != null)
+                        {
+                            foreach (var member in symbol.GetMembers())
+                            {
+                                switch (member.Kind)
+                                {
+                                    case SymbolKind.Namespace:
+                                    case SymbolKind.NamedType:
+                                        namespacesAndTypesToProcess.Push((NamespaceOrTypeSymbol)member);
+                                        break;
+                                    default:
+                                        throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                                }
+                            }
+                        }
+                        break;
+                    case SymbolKind.NamedType:
+                        var methodDocumentList = GetDocumentsForMethods(symbol, context, out var allMethodsHaveIL);
+
+                        // If all of the methods have IL then the document data we already have is enough, but if not
+                        // then we want to add document info for the type.
+                        if (!allMethodsHaveIL)
+                        {
+                            var typeDefinition = (Cci.ITypeDefinition)symbol.GetCciAdapter();
+
+                            foreach (var loc in symbol.Locations)
+                            {
+                                if (!loc.IsInSource)
+                                {
+                                    continue;
+                                }
+
+                                var span = loc.GetLineSpan();
+                                var debugDocument = DebugDocumentsBuilder.TryGetDebugDocument(span.Path, basePath: null);
+
+                                // We don't need to duplicate the data, if the method debug info would already contain this document
+                                if (debugDocument is not null && !methodDocumentList.Contains(debugDocument))
+                                {
+                                    result.Add(typeDefinition, debugDocument);
+                                }
+                            }
+                        }
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(symbol.Kind);
+                }
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Gets a list of documents from the method definitions in this type, and also outputs whether there are
+        /// any methods that don't have document info (ie don't have any IL)
+        /// </summary>
+        private static HashSet<Cci.DebugSourceDocument> GetDocumentsForMethods(NamespaceOrTypeSymbol typeSymbol, EmitContext context, out bool allMethodsHaveIL)
```

Comment:
Consider taking a `HashSet<T>` as a parameter and let the calling function amortize allocations with a `PooledHashSet<T>`. As implemented we are creating a `HashSet<T>` per typed considered here.